decomposer.go

```diff
@@ -69,7 +71,19 @@ func (d *Decimal) Decompose(buf []byte) (form byte, negative bool, coefficient [
 	// Finite form.
 	negative = d.Negative
 	exponent = d.Exponent
-	coefficient = d.Coeff.Bytes()
+
+	sizeInBytes := (d.Coeff.BitLen() + 8 - 1) / 8 // math.Ceil(d.Coeff.BitLen()/8.0)
+	if cap(buf)-len(buf) >= sizeInBytes {
+		// it extends the buffer as the filling of bytes expects an already
+		// allocated slice.
+		buf = append(buf, make([]byte, sizeInBytes)...)
```

Comment:
Have you verified that the `make([]byte, sizeInBytes)` does not lead to an actual heap allocation. If it does, we'd lose the benefit of this change.

Reply:
Yes that's highlighted by the benchmark:
```
BenchmarkDecomposerDecompose/no_allocation-8         	73670545	        14.98 ns/op	       0 B/op	       0 allocs/op
BenchmarkDecomposerDecompose/one_allocation-8        	42363051	        27.63 ns/op	       8 B/op	       1 allocs/op
```

Reply:
First one is providing a buffer and the other is providing nil.